Integrating multi-server switches and Libshard

bremanandjk opened the pull request:
**Introducing Multi-server switches:**
* As part of scalability work, we are improving the performance with a horizontal scalability model. This newly added multi-server switches **PBS_MAX_SERVERS** and **PBS_SERVER_INSTANCES** are used to enable running multiple PBS servers parallelly. 

* Implemented the code in a way to read the pbs.conf file and process it in pbs_loadconf, also added functions to get the values like get_max_servers() and get_currect_servers().

**Design Doc:** https://pbspro.atlassian.net/wiki/spaces/PD/pages/1585545237/Sharding+algorithm+design

**Introducing Libshard library:** 

* Since multiple servers are running, each client should know the whereabouts of the running server instances. To solve this problem, I have developed the shared library that every client would link and invoke the exported API's for finding a suitable server based on sharding policy and sharding algorithm(more details in the linked doc). 

* Namely libshard.so for Linux and Libshard.dll in windows. 

**Integration of Libshard in PBS commands:** 

* Have implemented the code changes to link the Libshard library in both Linux and Windows platforms.
* Each command in PBS Pro uses the get_svr_shard_connection() function to get the connected socket to communicate, the general eventually, even the Scheduler and the Mom would use this same function to get the TPP stream. Have taken considerable efforts to use the existing connection table for preserving the connected socket info with all servers. Added getters and setters for each newly added members in connection handle.
* This implementation includes changes in all IFL functions, that each IFL function would accept virtual socket and demux the actual physical socket dynamically with the help of the shard library.  

**Test results:** 
1) For GNU/Linux side implementation changes, we can refer to Travis and Jenkin results. 
https://github.com/bremanandjk/pbspro/runs/548702636

![image](https://user-images.githubusercontent.com/17980660/78034819-3aca8300-7336-11ea-92a8-132d8047cde5.png)

2) For windows side implementation, I have tested the windows JOBS from Windows client to Linux server. Logs are attached below.
a) https://ci.appveyor.com/project/bremanandjk/pbspro/builds/31848867
b) [Windows_command_job_submission.txt](https://github.com/subhasisb/pbspro/files/4409495/Windows_command_job_submission.txt)

**Test results for Multi-server mode:**

[Windows_command_job_submission.txt](https://github.com/subhasisb/pbspro/files/4454166/Windows_command_job_submission.txt)

bremanandjk commented:
**TH3 results:** 

![image](https://user-images.githubusercontent.com/17980660/78035934-b547d280-7337-11ea-990a-cc396f96444b.png)

Ran all the **regression** test cases, there are  3144 tests in total, out of that 3137 tests are pass. All the PTL test cases are success and 7 failures in Perl tests seems not stable. I will debug furthermore and confirm concretely that those failures are not directly related to these changes.

suresh-thelkar commented on src/include/Makefile.am:
```diff
@@ -41,7 +41,8 @@ include_HEADERS = \
 	tm_.h \
-	tm.h
+	tm.h \
+	libshard.h
```
Can we just say shard.h instead of libshard.h? This way it is consistent with other header files.

bremanandjk replied:
I initially kept this file name as "libshard.h", then changes to "shard_internal.h" as per your review comment, later got a comment from Subhasis to change back to "libshard.h", now changed it.

I am not inclined to any name, I am ok to change again, but let us confirm one name and I will finally change to that name.

suresh-thelkar replied:
I was actually asking to have two header files i.e. shard_internal.h for internal functions and shard.h for the actually library interface. I am fine if you want to leave this as libshard.h

suresh-thelkar commented on src/include/libpbs.h:
```diff
@@ -113,11 +113,25 @@ extern char pbs_current_group[];
+	int secondary_sd;           /* Secondary File descriptor for the open socket */
+	int state;                  /* Connection state */
+	time_t state_change_time;   /* Connnetion state change time */
+	time_t last_used;           /* Last used time for the connection */
```
Can we rename this to last_used_time?

bremanandjk replied:
done

suresh-thelkar commented on src/include/libpbs.h:
```diff
@@ -361,8 +379,9 @@ extern void pbs_authors(void);
 extern int PBSD_cred(int, char *, char *, int, char *, long, int, char **);
-
 int tcp_send_auth_req(int, unsigned int, char *);
+int tcp_send_auth_req(int sock, unsigned int port, char *user);
```
This prototype is same as the previous one. Why do we need it?

bremanandjk replied:
I have removed the duplicate entry.

suresh-thelkar commented on src/include/libshard.h:
```diff
@@ -0,0 +1,119 @@
+struct server_instance {
+    char *hostname;
+    int port;
+};
```
Can we use typedef for server_instance so that when we use this type at other places we no need to prefix it with the keyword struct.

bremanandjk replied:
done

suresh-thelkar commented on src/lib/Libdb/db_postgres_job.c:
```diff
@@ -727,3 +730,34 @@ pg_db_reset_job(pbs_db_obj_info_t *obj)
+
+	PQclear(res);
+
+	*jobid = last_jobid;
```
Do a null check for jobid.

bremanandjk replied:
Done

suresh-thelkar commented on src/lib/Libifl/conn_table.c:
```diff
@@ -520,3 +520,145 @@ get_conn_mutex(int fd)
+ * @par MT-safe: Yes
+ */
+int
+get_conn_shard_context(int fd)
```
It is better to name fd as virtual fd i.e. vfd as you have done with the other function for consistency.

bremanandjk replied:
Done

suresh-thelkar commented on src/lib/Libifl/conn_table.c:
```diff
@@ -520,3 +520,145 @@ get_conn_mutex(int fd)
+ * @par MT-safe: Yes
+ */
+int
+set_conn_shards(int fd, struct shard_conn ** shards)
```
Here also you can name fd as vfd.

suresh-thelkar replied:
As I have mentioned earlier we also need one more setter method as given below.
int set_conn_shard_at_index(int vfd, shard_conn *shard, int index)
This will basically set the given shard entry at the given index. If we don't have this routine the caller of this function always need to send entire shard_conn array every time even though the caller wants to modify a single entry in the shard_conn table.

bremanandjk replied:
For PBS commands integration, I don't see any direct use-case, 
If needed in the future, we will add that function. what do you say?

suresh-thelkar replied:
As discussed the use case is MOM and Scheduler integration with libshard. As Scheduler/MOM gets server connections one at at time this API becomes handy that time.

suresh-thelkar commented on src/lib/Libifl/int_hook.c:
```diff
@@ -99,12 +102,12 @@ PBSD_hookbuf(int c, int reqtype, int seq, char *buf, int len, char *hook_filenam
 	}
 
-	if (dis_flush(c)) {
+	if (dis_flush(sock)) {
```
As we discussed we can use call to dis_flush() for both TPP and TCP.

bremanandjk replied:
Done in new commit.

suresh-thelkar commented on src/lib/Libifl/int_submit.c:
```diff
@@ -515,23 +524,30 @@ char *
-	char *return_jobid = NULL;
-	int rc;
+	char  *return_jobid = NULL;
+	int    rc;
```
Just put one space after datatype that way it is consistent with the others.

bremanandjk replied:
Done

suresh-thelkar commented on src/lib/Libifl/pbsD_connect.c:
```diff
@@ -253,6 +256,288 @@ get_hostsockaddr(char *host, struct sockaddr_in *sap)
+			setenv("SystemRoot", "C:\\WINDOWS", 1);
+		}
+#endif
+	sd = socket(AF_INET, SOCK_STREAM, 0);	
```
Check the return value of socket system call.

bremanandjk replied:
Done

suresh-thelkar commented on src/lib/Libifl/pbsD_connect.c:
```diff
@@ -253,6 +256,288 @@ get_hostsockaddr(char *host, struct sockaddr_in *sap)
+#endif
+	sd = socket(AF_INET, SOCK_STREAM, 0);	
+
+	strcpy(pbs_server, server); /* set for error messages from commands */
```
Use strncpy.

bremanandjk replied:
Done

suresh-thelkar commented on src/lib/Libifl/pbsD_connect.c:
```diff
@@ -253,6 +256,288 @@ get_hostsockaddr(char *host, struct sockaddr_in *sap)
+ * @retval -1	error encountered setting up the connection.
+ */
+
+int internal_connect_cli(int vsock, char *server, int port, char *extend_data)
```
Can we rename this as internal_client_connect?

bremanandjk replied:
I felt all the Commands, Scheduler and MoM as clients, but this current name is specific to cli(command-line interface). It should be fine to keep this name I think.

suresh-thelkar replied:
I was just saying expand the abbreviation cli to client and make it internal_client_connect which is more readable.

bremanandjk replied:
I have renamed as per your suggestion.

suresh-thelkar commented on src/lib/Libifl/pbsD_connect.c:
```diff
@@ -253,6 +256,288 @@ get_hostsockaddr(char *host, struct sockaddr_in *sap)
+	static int shard_init_flag = -1;
+	int *inactive_servers; 
+	int num_of_conf_servers = get_current_servers();
+	struct shard_conn **shard_connection = get_conn_shards(vsock);
```
Do a null check for shard_connection.

bremanandjk replied:
Have added check accordingly.

suresh-thelkar commented on src/lib/Libifl/pbsD_connect.c:
```diff
@@ -253,6 +256,288 @@ get_hostsockaddr(char *host, struct sockaddr_in *sap)
+
+	if (get_max_servers() > 1) {
+		if (shard_init_flag == -1) {
+			if (pbs_shard_init(get_max_servers(), (struct server_instance **)pbs_conf.psi, num_of_conf_servers) == -1)
```
Better not to call get_max_servers() multiple times. Instead collect the output of get_max_servers() and use it wherever needed.

bremanandjk replied:
Done

suresh-thelkar commented on src/lib/Libifl/pbsD_connect.c:
```diff
@@ -253,6 +256,288 @@ get_hostsockaddr(char *host, struct sockaddr_in *sap)
+	int i = 0;
+	int inact_srv_index = 0;
+	static int shard_init_flag = -1;
+	int *inactive_servers; 
```
Can we rename this as inact_svr_indexes?

bremanandjk replied:
I have renamed it.

suresh-thelkar commented on src/lib/Libifl/pbsD_connect.c:
```diff
@@ -253,6 +256,288 @@ get_hostsockaddr(char *host, struct sockaddr_in *sap)
+		} else if (shard_connection[srv_index]->state == SHARD_CONN_STATE_CONNECTED) {
+			return  shard_connection[srv_index]->sd;
+		}
+	set_conn_shards(vsock, shard_connection);
```
Any reason why we are not maintaining conn_exists flag and use it here which we used to do earlier. If connection exists initially connection state might be down. In the current code it tries to get a new connection if it is down, this is something we don't want when Scheduler uses libshard.

bremanandjk replied:
For PBS command integration, "conn_exists" flag is not required, thus I have not integrated it now.

suresh-thelkar replied:
One of the use cases is Scheduler has existing connections and this flag will be useful there.

suresh-thelkar commented on src/lib/Libifl/pbsD_connect.c:
```diff
@@ -305,6 +588,33 @@ __pbs_connect_extend(char *server, char *extend_data)
+#endif
+	sd = socket(AF_INET, SOCK_STREAM, 0);	
+
+	if ((strcmp(server,pbs_default()) == 0) && pbs_conf.pbs_max_servers > 1) {
```
use get_max_servers() function here.

bremanandjk replied:
done

suresh-thelkar commented on src/lib/Libifl/pbsD_connect.c:
```diff
@@ -305,6 +588,33 @@ __pbs_connect_extend(char *server, char *extend_data)
+			shard_connection[i]->state_change_time = 0;
+			shard_connection[i]->last_used = 0;
+		}
+		set_conn_shards(sd, shard_connection);
```
Line No 602 to 611 can be wrapped inside a function like initialise_shard_conn() so that this is can be reused at Scheduler and MOM side as well.

bremanandjk replied:
Done in the new commit.

suresh-thelkar commented on src/lib/Libifl/pbsD_connect.c:
```diff
@@ -305,6 +588,33 @@ __pbs_connect_extend(char *server, char *extend_data)
+
+
+		/* can't use failover for now, take a different path */
+	return sd; /* actual connects happen when client tries to send a request first */
```
Can we align the commented strings and also add more description towards failover.

bremanandjk replied:
Done

suresh-thelkar commented on src/lib/Libifl/pbsD_connect.c:
```diff
@@ -336,61 +646,19 @@ __pbs_connect_extend(char *server, char *extend_data)
-			return -1; /* pbs_errno was set */
-	}
-
 	/*
```
Any reason why this code is removed?

suresh-thelkar replied:
Brem, Can you please look into this question?

bremanandjk replied:
I have added this code, It was missed by mistake.

suresh-thelkar commented on src/lib/Libifl/pbsD_defschreply.c:
```diff
@@ -87,18 +88,25 @@ pbs_defschreply(int c, int cmd, char *id, int err, char *txt, char *extend)
 		return pbs_errno;
+		
+	set_new_shard_context(c);
+	sock = get_svr_shard_connection(c, PBS_BATCH_DefSchReply, id);
```
It looks we are passing PBS_BATCH_* as req_type in some of the places where we call get_svr_shard_connection even though that is not getting used inside this function. This info will be used for taking better decisions in future. So can we make sure we pass req_type in all the places when we call this function.

bremanandjk replied:
I have added the necessary BATCH request types in each for its call.

suresh-thelkar commented on src/lib/Libifl/pbsD_termin.c:
```diff
@@ -64,8 +123,11 @@
+	int i;
+	int sock;
+	int count = 0;
+	struct shard_conn **shard_connection = get_conn_shards(c);
```
Do a null check for shard_connection.

bremanandjk replied:
Done

suresh-thelkar commented on src/lib/Libifl/pbs_loadconf.c:
```diff
@@ -96,6 +96,9 @@ struct pbs_config pbs_conf = {
 	NULL,					/* pbs_server_name */
 	NULL,					/* PBS server id */
+	1,						/* single pbs server instance by default */
+	0,						/* single pbs server instance by default */
```
Can we align both of these commented strings with the other ones for consistency.

bremanandjk replied:
Done,

suresh-thelkar commented on src/lib/Libifl/pbs_loadconf.c:
```diff
@@ -252,6 +255,121 @@ parse_config_line(FILE *fp, char **key, char **val)
+			fprintf(stderr, "Ran out of memory parsing configuration %s", conf_value);
+			return -1;
+		}
+		pbs_conf.psi[count]->name = strdup(host);
```
check the return value of strdup() and use accordingly.

bremanandjk replied:
Have checked it

suresh-thelkar commented on src/lib/Libifl/pbs_loadconf.c:
```diff
@@ -252,6 +255,121 @@ parse_config_line(FILE *fp, char **key, char **val)
+
+	/* ignore any PBS_SERVER_INSTANCES and default fill */
+	if (pbs_conf.psi) {
+		for(i = 0; i < pbs_conf.pbs_current_servers; i++) {
```
use get_current_servers() instead.

bremanandjk replied:
Sure, added code accordingly.

suresh-thelkar commented on src/lib/Libshard/Makefile.am:
```diff
@@ -0,0 +1,48 @@
+
+libshard_la_CPPFLAGS = -I$(top_srcdir)/src/include
+
+libshard_la_LDFLAGS = -version-info 0:0:0
```
Can we have version as 1.0.0

bremanandjk replied:
done

suresh-thelkar commented on src/lib/Libshard/shard_functions.c:
```diff
@@ -0,0 +1,224 @@
+#define DLLEXPORT
+#endif
+
+int compute_srv_ind(char *obj_id);
```
Can we have this in a header file?

bremanandjk replied:
It's local to this file, so not added in header.

suresh-thelkar commented on src/lib/Libshard/shard_functions.c:
```diff
@@ -0,0 +1,224 @@
+{
+	// Note: some broken versions only have 8 trailing zero's, the correct epoch has 9 trailing zero's
+	// This magic number is the number of 100 nanosecond intervals since January 1, 1601 (UTC)
+	// until 00:00:00 January 1, 1970 
```
Don't use c++ style comments.

bremanandjk replied:
ok, done

suresh-thelkar commented on src/lib/Libutil/misc_utils.c:
```diff
@@ -2158,3 +2158,43 @@ crc_file(char *filepath)
+	return pbs_conf.pbs_current_servers;
+}
+
+int
```
Write Doxygen header.

bremanandjk replied:
Added Doxygen header.

suresh-thelkar commented on src/lib/Libutil/misc_utils.c:
```diff
@@ -2158,3 +2158,43 @@ crc_file(char *filepath)
+		for(i = 0; i < get_current_servers(); i++) {
+			if (instance->port == pbs_conf.psi[i]->port) {
+				if (strcmp(instance->name, pbs_conf.psi[i]->name) == 0) 
+					return i;
```
Have a NULL check for instance->name for pbs_conf.psi[i]->name or else we have to mandate both host:port while configuring psi.

bremanandjk replied:
Done with this suggestion.

suresh-thelkar commented on src/server/pbsd_init.c:
```diff
@@ -567,7 +569,20 @@ pbsd_init(int type)
+			svr_jobidnumber = njobid;
+			if (svr_jobidnumber == -1) {
+				log_err(-1, __func__, "Failed to compute svr_jobidnumber");
+				return (-1);
```
We are not computing anything here apart from getting the desired jobid from database. Can we have better log message?

bremanandjk replied:
I have addressed this comment.

suresh-thelkar commented on src/server/pbsd_main.c:
```diff
@@ -833,6 +836,9 @@ main(int argc, char **argv)
 	if (pbs_loadconf(0) == 0)
 		return (1);
+	/* initialize the shard lib */
+	if (pbs_shard_init(get_max_servers(), (struct server_instance **)pbs_conf.psi, get_current_servers()) == -1)
```
can we call either get_max_servers() or get_current_servers() once, collect the output and use it instead of calling it multiple times.

bremanandjk replied:
Sure, done

suresh-thelkar commented on src/server/pbsd_main.c:
```diff
@@ -864,6 +870,16 @@ main(int argc, char **argv)
 
+	self.name = strdup(server_host);
+	self.port = pbs_conf.batch_service_port;
+		if (get_max_servers() > 1) {
```
Please align these statements with the above.

bremanandjk replied:
Done

suresh-thelkar commented on src/lib/Libifl/conn_table.c:
```diff
@@ -520,3 +520,145 @@ get_conn_mutex(int fd)
+		UNLOCK_TABLE(-1);
+		return -1;
+	}
+	p->ch_shards = shards;
```
This function is just taking a reference to ch_shards table. It means that we are expecting clients to create this table. This allows clients to accidentally modify ch_shards table without using the proper getters/setter methods. If we want to go with the proper encapsulation then it is always better to create this ch_shards table inside conn_table.c or library in a static method and only expose getters and setters for the same. This is more safer and allows us to achieve encapsulation and abstraction. What do you think?

bremanandjk replied:
Added code to allocate shard_connection within conn_table file. so that encapsulation is achieved.

bremanandjk commented:
@suresh-thelkar Thanks for your reviews, 
I have addressed most of your comments and replied to some. 

Please have a look and review further changes, Thanks.

suresh-thelkar commented on src/lib/Libifl/conn_table.c:
```diff
@@ -627,6 +627,22 @@ set_conn_shards(int fd, struct shard_conn ** shards)
 
+int initialise_shard_conn(int vfd){
+	int i;
+	struct shard_conn **shard_connection = calloc(get_current_servers(), sizeof(struct shard_conn *));
```
No need to call get_current_servers() multiple times. Call it initially, collect its return value and use it further.

bremanandjk replied:
Done

suresh-thelkar commented on src/lib/Libifl/conn_table.c:
```diff
@@ -627,6 +627,22 @@ set_conn_shards(int fd, struct shard_conn ** shards)
 	return 0;
 }
 
+int initialise_shard_conn(int vfd){
```
Please write Doxygen header for this function.

suresh-thelkar commented on src/lib/Libifl/int_hook.c:
```diff
@@ -97,17 +100,16 @@ PBSD_hookbuf(int c, int reqtype, int seq, char *buf, int len, char *hook_filenam
 	}
 
+	if (prot == PROT_TPP) {
+		return pbs_errno;
```
Do we need this return statement as we are returning pbs_errno just prior to this.

bremanandjk replied:
This return is needed for TPP. Even in the mainline, it returns pbs_errno.

bremanandjk replied:
The reason being, In the case of TPP we don't want to read the reply after dis_flush(). 
If the dis_flush() fails, we are returning the pbs_errno as PBSE_PROTOCOL. This block is common to both TCP and TPP protocols.

If dis_flush() succeeds for the TPP protocol, we are returning with the PBSE_NONE error code.

you could refer the same in mainline, https://github.com/PBSPro/pbspro/blob/master/src/lib/Libifl/int_hook.c#L100-L105

suresh-thelkar commented on src/lib/Libifl/int_jcred.c:
```diff
@@ -77,34 +77,34 @@ PBSD_jcred(int c, int type, char *buf, int len, int prot, char **msgid)
-	if (dis_flush(c)) {
-		return (pbs_errno = PBSE_PROTOCOL);
+	if (prot == PROT_TPP) {
+		return pbs_errno;
```
Irrespective of either the protocol is TPP or not we are returning pbs_errno in case of failure above. Then why do we need to return pbs_errno here again ?. This always returns pbs_errno if protocol is TPP even in case of dis_flush() success.

bremanandjk replied:
It's an intentional return.

suresh-thelkar commented on src/server/pbsd_main.c:
```diff
@@ -864,6 +870,16 @@ main(int argc, char **argv)
+	self.port = pbs_conf.batch_service_port;
+	if (get_max_servers() > 1) {
+		if ((myindex = get_svr_index(&self)) == -1) {
+			fprintf(stderr, "Wrong Multi Server configuration. Please start server after correcting /etc/pbs.conf\n");
```
Do we need to add this to the design document as this is external message ?

bremanandjk replied:
I have added this interface in the Design doc. Thanks

subhasisb commented on src/server/pbsd_main.c:
```diff
@@ -872,12 +872,12 @@ main(int argc, char **argv)
-			}
+	if (get_max_servers() > 1) {
+		if ((myindex = get_svr_index(&self)) == -1) {
+			fprintf(stderr, "Wrong Multi Server configuration. Please start server after correcting /etc/pbs.conf\n");
```
pbs.conf may not be in /etc/pbs.conf, also think of windows. Please give consistent error messages comparing with other such error message in pbs_loadconf

bremanandjk replied:
Added error msg accordingly.

bremanandjk commented on src/lib/Libifl/conn_table.c:
```diff
@@ -627,6 +627,22 @@ set_conn_shards(int fd, struct shard_conn ** shards)
 	return 0;
 }
 
+int initialise_shard_conn(int vfd){
```
I have added headers to this function in the following commit.

bremanandjk commented on src/lib/Libifl/conn_table.c:
```diff
@@ -643,21 +659,27 @@ set_conn_shards(int fd, struct shard_conn ** shards)
 		return NULL;
 	}
+	if (ch_shard_init == -1) {
+		if(initialise_shard_conn(vfd) == -1)
```
Fixed it.

bremanandjk commented on src/lib/Libifl/pbsD_connect.c:
```diff
@@ -454,30 +454,31 @@ get_svr_shard_connection(int vsock, int req_type, void *shard_hint)
 
-	if (get_max_servers() > 1) {
+	if (max_num_servers > 1) {
 		if (shard_hint || get_conn_shard_context(vsock) == -1 ) {
```
Thanks for your suggestion, I have done the same in the next commit.

bremanandjk commented on src/lib/Libifl/pbsD_connect.c:
```diff
@@ -490,7 +491,7 @@ get_svr_shard_connection(int vsock, int req_type, void *shard_hint)
 		if (inact_srv_index) {	
 			if (inact_srv_index == num_of_conf_servers)
```
I agree with this name change. I have addressed it.

bremanandjk commented on src/lib/Libifl/pbsD_connect.c:
```diff
@@ -454,30 +454,31 @@ get_svr_shard_connection(int vsock, int req_type, void *shard_hint)
 	int num_of_conf_servers = get_current_servers();
+	int max_num_servers = get_max_servers();
 	struct shard_conn **shard_connection = get_conn_shards(vsock);
 
```
Added check accordingly.

bremanandjk commented on src/lib/Libifl/pbsD_connect.c:
```diff
@@ -490,7 +491,7 @@ get_svr_shard_connection(int vsock, int req_type, void *shard_hint)
 			if (inact_srv_index == num_of_conf_servers)
 				return -1;
-			srv_index = pbs_shard_get_server_byindex(shard_hint, JOB, inactive_servers);
+			srv_index = pbs_shard_get_server_byindex(shard_hint, JOB, inact_svr_indexes);
```
Yes, it is stored in the libshard library.

bremanandjk commented on src/lib/Libifl/pbs_loadconf.c:
```diff
@@ -320,7 +320,10 @@ parse_psi(char * conf_value)
 		}
-		pbs_conf.psi[count]->name = strdup(host);
+		if (!(pbs_conf.psi[count]->name = strdup(host))) {
+			fprintf(stderr, "Ran out of memory");
```
Added full sentence.

bremanandjk commented on src/lib/Libshard/shard_functions.c:
```diff
@@ -67,15 +67,15 @@
-	// Note: some broken versions only have 8 trailing zero's, the correct epoch has 9 trailing zero's
-	// This magic number is the number of 100 nanosecond intervals since January 1, 1601 (UTC)
-	// until 00:00:00 January 1, 1970 
+	/* Note: some broken versions only have 8 trailing zero's, the correct epoch has 9 trailing zero's
```
Here, it is not possible to add this function in the Libwin library due to the fact, there is inter-dependency between Libwin and Libpbs. And Libpbs depends on Libshard. 

This creates a circular dependency between these three libraries.

Libpbs <--> Libshard <--> Libwin  <--> Libpbs

subhasisb replied:
ok we will leave this for later cleanup of libraries

bremanandjk commented:
Thanks, @subhasisb and @suresh-thelkar for your review comments, 

I have addressed all your review comments and attached test results for Multi-server mode. 

Please share your further reviews. 

P.S: All the TH3 regression test cases are passed. 

![image](https://user-images.githubusercontent.com/17980660/78860603-9f55a400-7a00-11ea-9c6b-8da596ae3e2a.png)

subhasisb commented on src/server/pbsd_main.c:
```diff
@@ -874,7 +874,7 @@ main(int argc, char **argv)
 	if (get_max_servers() > 1) {
 		if ((myindex = get_svr_index(&self)) == -1) {
-			fprintf(stderr, "Wrong Multi Server configuration. Please start server after correcting /etc/pbs.conf\n");
+			fprintf(stderr, "pbsconf error: Wrong Multi Server configuration.\n");
```
Dont add a period at the end of the message - we do not put a period at the end of any messages

bremanandjk replied:
Done, Thanks for letting me know this.

subhasisb commented on src/lib/Libifl/pbs_loadconf.c:
```diff
@@ -321,7 +321,7 @@ parse_psi(char * conf_value)
 		}
 		if (!(pbs_conf.psi[count]->name = strdup(host))) {
-			fprintf(stderr, "Ran out of memory");
+			fprintf(stderr, "Ran out of memory parsing multi-server configuration");
```
Don't you need a \n here?

bremanandjk replied:
I have added \n here.

subhasisb commented on src/lib/Libifl/pbsD_connect.c:
```diff
@@ -467,39 +471,41 @@ get_svr_shard_connection(int vsock, int req_type, void *shard_hint)
 	}
 	if (internal_connect == NULL)
-		internal_connect = internal_connect_cli;
+		internal_connect = internal_client_connect;
```
Better to rename internal_connect to "pfn_connect" or something similar

bremanandjk replied:
pfn_connect looks fine for me, addressed it.

bremanandjk commented on src/lib/Libifl/conn_table.c:
```diff
@@ -627,12 +627,23 @@ set_conn_shards(int vfd, struct shard_conn ** shards)
+ * @retval 0 - success
+ * @retval -1 - error
+ */
 int initialise_shard_conn(int vfd){
```
I have addressed it.

bremanandjk commented on src/lib/Libifl/pbsD_connect.c:
```diff
@@ -292,9 +292,13 @@ internal_tcp_connect(int vsock, char *server, int server_port, char *extend_data
+	}	
 
-	strcpy(pbs_server, server); /* set for error messages from commands */
+	strncpy(pbs_server, server, sizeof(pbs_server)-1); /* set for error messages from commands */
```
Added a line for NULL termination.

bremanandjk commented on src/server/pbsd_init.c:
```diff
@@ -567,7 +569,19 @@ pbsd_init(int type)
 		 * see: get_next_svr_sequence_id(void)
 		 */
-		svr_jobidnumber = server.sv_qs.sv_jobidnumber;
+		if (pbs_db_get_maxjobid(conn, &njobid) == -1) {
```
I have removed this DB related code. Thanks for catching this.

bremanandjk commented on src/lib/Libdb/db_postgres_job.c:
```diff
@@ -727,3 +730,37 @@ pg_db_reset_job(pbs_db_obj_info_t *obj)
 	free_db_attr_list(&(obj->pbs_db_un.pbs_db_job->attr_list));
 }
 
+
```
Sure, I have addressed it.

bremanandjk commented on src/include/libshard.h:
```diff
@@ -0,0 +1,119 @@
+typedef struct server_instance {
+    char *hostname;
+    int port;
+}server_instance_t;
```
Done.

bremanandjk commented on src/lib/Libifl/conn_table.c:
```diff
@@ -520,3 +520,178 @@ get_conn_mutex(int fd)
+ * @retval 0 - success
+ * @retval -1 - error
+ */
+int initialise_shard_conn(int vfd){
```
ok, done

bremanandjk commented:
@subhasisb @suresh-thelkar 
Thanks for your detailed code reviews and I have addressed them all. Also, I have answered if any questions asked in reviews. Please share further comments if any! 

 Regression results:

![image](https://user-images.githubusercontent.com/17980660/79033087-5d03a280-7b79-11ea-8cae-4e8e7136868f.png)

suresh-thelkar commented:
@subhasisb, As there are some code changes in extensions repo also for this PR, I request you to create an integration branch for Multi Server in extensions.

bremanandjk commented:
Thanks @suresh-thelkar for your reviews, 

I agree with your design proposal, yes it would be better to make it purely encapsulation for connection structure. However, while observing the getter functions for other members in this connection structure like get_conn_chan(), get_conn_mutex(), get_conn_errtxt(), they are all returning pointers. (not const data or return by value) 

If we want to achieve full abstraction in connection structure, we would need to modify all of them. I feel we could do that changes separately for all the getters.

suresh-thelkar commented:
> Thanks @suresh-thelkar for your reviews,
> 
> I agree with your design proposal, yes it would be better to make it purely encapsulation for connection structure. However, while observing the getter functions for other members in this connection structure like get_conn_chan(), get_conn_mutex(), get_conn_errtxt(), they are all returning pointers. (not const data or return by value)
> 
> If we want to achieve full abstraction in connection structure, we would need to modify all of them. I feel we could do that changes separately for all the getters.

@bremanandjk, Thanks for agreeing with me. Let us take @subhasisb thoughts on the same and then we can proceed accordingly.

hirenvadalia commented:
> However, while observing the getter functions for other members in this connection structure like get_conn_chan(), get_conn_mutex(), get_conn_errtxt(), they are all returning pointers. (not const data or return by value)

(Maybe I am missing here) but the application never uses directly a member of the TCP connection structure (pbs_tcp_chan_t, which is returned by get_conn_chan), we have getter/setter APIs (defined in dis_helpers.c, for example, transport_chan_set_ctx_status() etc..) to get/set value inside pbs_tcp_chan_t struct,

+1 on @suresh-thelkar proposal.

bremanandjk commented:
Here, we are not talking about how today's code in the application layer is using the getters in connection structure. In order to make sure the complete abstraction, and to have restrictions to posterity developers by direct access to internal members, we should not return raw pointers at all. If we expose pointers, always there is a feasibility to access internal members. Hence encapsulation is not fully achieved here.

subhasisb commented:
ok here is what i think. The shard_connection structure is needed in the application. pbs_connect() and other IFL calls will call get_svr_shard_connection() which needs to access things inside shard_connection structure. Thus access to those is required. What is not required is for conn_table.c to know the structure of shard_connection. For the connection table, this can simply be pointer to "application" data designated by a name say "shard_conn_data" of void * type.

U can have one setter and getter so that the application can create the shard connection array and set it  via the setter and return it via the getter.

subhasisb commented on src/lib/Libifl/int_ucred.c:
```diff
@@ -82,7 +82,7 @@ PBSD_ucred(int c, char *user, int type, char *buf, int len)
 	int	sock;
 
-	sock = get_svr_shard_connection(c, -1, NULL);
+	sock = get_svr_shard_connection(c, PBS_BATCH_UserCred, NULL);
```
PBS_BATCH_UserCred is being removed by the windows team, so this will be gone

bremanandjk replied:
ok

subhasisb commented on src/lib/Libifl/int_ucred.c:
```diff
@@ -144,7 +144,7 @@ PBSD_user_migrate(int c, char *tohost)
 	int	sock;
 
-	sock = get_svr_shard_connection(c, -1, NULL);
+	sock = get_svr_shard_connection(c, PBS_BATCH_UserMigrate, NULL);
```
PBS_BATCH_UserMigrate is being removed by the windows team and so we can forget about this one as well

subhasisb commented on src/lib/Libifl/pbsD_stagein.c:
```diff
@@ -75,7 +75,7 @@ pbs_stagein(int c, char *jobid, char *location, char *extend)
 		location = "";
 
-	sock = get_svr_shard_connection(c, -1, NULL);
+	sock = get_svr_shard_connection(c, PBS_BATCH_StageIn, NULL);
```
This should pass jobid! 

Also, pbs_stagein does not seem to be a top level API , and neither is any code using it?

bremanandjk replied:
No reference anywhere, let me remove it in the next commit.

bremanandjk commented on src/lib/Libifl/int_jcred.c:
```diff
@@ -80,7 +80,7 @@ PBSD_jcred(int c, int type, char *buf, int len, int prot, char **msgid)
 
 	if (prot == PROT_TCP) {
-		sock = get_svr_shard_connection(c, -1, NULL);
+		sock = get_svr_shard_connection(c, PBS_BATCH_JobCred, NULL);
```
I have retained the batch id for future use if we decide to use sharding based on it. Some of the references are using batch id, for consistency purpose I have added in all other remaining places.

bremanandjk commented on src/lib/Libifl/int_msg2.c:
```diff
@@ -74,7 +74,7 @@ PBSD_msg_put(int c, char *jobid, int fileopt, char *msg, char *extend, int prot,
 
 	if (prot == PROT_TCP) {
-		sock = get_svr_shard_connection(c, -1, NULL);
+		sock = get_svr_shard_connection(c, PBS_BATCH_MessJob, NULL);
```
`> And yet you are not passing the jobid to it? U are passing NULL!!! It would choose a random server and go there! `

This is not true always. Only if the shard_context is reset to "-1", then passing NULL as jobid would choose the random server. Else get_svr_shard_connection() would refer the shard_context and return the last used server's socket in case of dialogue. 

For PBSD_msg_put(), the server connection happens in its parent API "__pbs_msgjob()". There we are passing the jobid after reset and preserve the server index in the connection table. 
The internal functions namely 1) PBSD_msg_put() and 2)PBSD_rdrpy() would use the last used server. 

Please let me know if I am wrong.

bremanandjk commented on src/lib/Libifl/int_msg2.c:
```diff
@@ -152,7 +152,7 @@ PBSD_relnodes_put(int c, char *jobid, char *node_list, char *extend, int prot, c
 
 	if (prot == PROT_TCP) {
-		sock = get_svr_shard_connection(c, -1, NULL);
+		sock = get_svr_shard_connection(c, PBS_BATCH_RelnodesJob, NULL);
```
An actual connection happens at parent API pbs_relnodesjob() and its internal functions would use an already connected server. To get the last used server socket, we need to call get_svr_shard_connection() with a NULL value and without resetting the shard_context.

bremanandjk commented on src/lib/Libifl/int_submit.c:
```diff
@@ -249,7 +249,7 @@ PBSD_commit(int c, char *jobid, int prot, char **msgid)
 
 	if (prot == PROT_TCP) {
-		sock = get_svr_shard_connection(c, -1, NULL);
+		sock = get_svr_shard_connection(c, PBS_BATCH_Commit, NULL);
```
If my understanding is right, the current code does the same as you explain. The connection happens to the server once at top level (PBSD_queuejob) and all subsequent API's like PBSD_jscript() & PBSD_commit() would connect to the same server since the shard_context is untouched.

bremanandjk commented on src/lib/Libifl/int_submit.c:
```diff
@@ -191,7 +191,7 @@ PBSD_rdytocmt(int c, char *jobid, int prot, char **msgid)
 
 	if (prot == PROT_TCP) {
-		sock = get_svr_shard_connection(c, -1, NULL);
+		sock = get_svr_shard_connection(c, PBS_BATCH_RdytoCommit, NULL);
```
I see a usage in send_job() function and also there is a text in server code saying "The PBS_BATCH_Rdytocommit message is deprecated"  
I will remove it in the next commit after running regressions.

bremanandjk commented:
I have pushed a new commit after TH3 pass, in this I have addressed making the shard_connection opaque to conn_table.c

![image](https://user-images.githubusercontent.com/17980660/79174073-8dd91700-7dc7-11ea-95e3-355a96a0cd70.png)

bremanandjk commented:
We could not run this branch's PBS commands with a multi-server branch's PBS servers due to DIS protocol errors and difference in authentication batch requests.  If there are any possibilities, please let me know, I will try to run in actual multi-server environment and share the results. 

Still, we could prove that in pbs_submit(), all the internal API's namely a) pbs_queuejob(), b) pbs_jsscript(). and c) pbs_commit() connects to the same server using GDB. 

Qsub File: 
[Multi-Server_Job_Submission_GDB.txt](https://github.com/subhasisb/pbspro/files/4472555/Multi-Server_Job_Submission_GDB.txt)

qmsg: 1) PBSD_msg_put() and 2)PBSD_rdrpy()  connects to same server without referring shardlib. 
[Multi-Server_Job_Submission_GDB_qmsg.txt](https://github.com/subhasisb/pbspro/files/4472565/Multi-Server_Job_Submission_GDB_qmsg.txt)

Please help me, if I am missing anything. Thanks
